File: 프로그래머스/2/159993. 미로 탈출/미로 탈출.py
from collections import deque 

def bfs(maps, sx, sy, tx, ty):
    q = deque()
    visited = [[0] * len(maps[0]) for _ in range(len(maps))]
    q.append((sx, sy, 0))
    
    dx = [-1, 1, 0, 0]
    dy = [0, 0, -1, 1]
    
    while q:
        x, y, cnt = q.popleft()
        if x == tx and y == ty:
            return cnt 
        for i in range(4):
            nx, ny = x + dx[i], y + dy[i]
            if 0 <= nx < len(maps) and 0 <= ny < len(maps[0]) and visited[nx][ny] == 0 and maps[nx][ny] != 'X':
                q.append((nx, ny, cnt+ 1))
                visited[nx][ny] = 1
    return -1
                
                
    
    
    
def solution(maps):
    for i in range(len(maps)):
        for k in range(len(maps[0])):
            if maps[i][k] == 'S':
                sx, sy = i, k
            elif maps[i][k] == 'L':
                lx, ly = i, k
            elif maps[i][k] == 'E':
                ex, ey = i, k
                
    p1 = bfs(maps, sx, sy, lx, ly)
    p2 = bfs(maps, lx, ly, ex, ey)
    if p1 != -1 and p2 != -1:
        return p1+p2
    else:
        return -1



# 갔던 길을 또방문하는 최단 거리의 경우는 레버를 찾기 이전과 찾고 난 후 사이에만 존재할 수 있음 
# 두가지 과정으로 나누어 접근 

File: 프로그래머스/2/159993. 미로 탈출/미로 탈출.java
import java.util.LinkedList;
import java.util.Queue;


class Solution {
    
    static String[][] maze;
    static int[] dx = {-1, 1, 0, 0};
    static int[] dy = {0, 0, -1, 1};
    
    public int solution(String[] maps) {
        maze = new String[maps.length][maps[0].length()];
        int[] start = new int[2];
        int[] lever = new int[2];
        
        for (int i = 0 ; i < maps.length; i++) {
            String[] tmp = maps[i].split("");
            
            for(int k = 0 ; k < tmp.length; k++) {
                maze[i][k] = tmp[k];
                
                if (maze[i][k].equals("S")) {
                    start = new int[]{i, k};
                }
                
                if (maze[i][k].equals("L")) {
                    lever = new int[] {i, k};
                }
            }
        }
        int result1 = bfs(start, "L");
        int result2 = bfs(lever, "E");
        
        if (result1 == -1 || result2 == -1) {
            return -1;
        }
        return result1 + result2;
        
    }
    
    public int bfs(int[] start, String target) {
        Queue<int[]> q = new LinkedList<>();
        q.add(new int[]{start[0], start[1], 0});
        
        boolean[][] visited = new boolean[maze.length][maze[0].length];
        
        while(!q.isEmpty()) {
            int x = q.peek()[0];
            int y = q.peek()[1];
            int count = q.peek()[2];
            
            visited[x][y] = true;
            
            if (maze[x][y].equals(target)) {
                return count;
            }
            
            q.poll();
            for (int i = 0; i< 4; i++) {
                int nx = x + dx[i];
                int ny = y + dy[i];
                
                if (nx >= 0 && nx < maze.length && ny >= 0 && ny < maze[0].length && !visited[nx][ny])
                {
                    if (!maze[nx][ny].equals("X")) {
                        visited[nx][ny] = true;
                        q.add(new int[] {nx, ny, count + 1});
                    }
                    
                }            
            }
        }
        return -1;
    }
}